heudiconv/dicoms.py

```diff
@@ -314,19 +315,55 @@ def group_dicoms_into_seqinfos(files, grouping, file_filter=None,
     return seqinfos
 
 
-def get_dicom_series_time(dicom_list):
-    """Get time in seconds since epoch from dicom series date and time
-    Primarily to be used for reproducible time stamping
+def get_timestamp_from_series(dicom_list: list[str]) -> int:
+    """try to return a timestamp indicating when the first dicom in dicom_list was created, or if that
+    info isn't present then a reproducible integer. This is used in setting mtimes reproducibly
+
+    Args:
+        dicom_list (list[str]): list of strings pointing to existing dicom files
+
+    Returns:
+        int: either an int representing when the first dicom was created, 
+            in number of seconds since epoch, or if no datetime info is found then a hash of the
+            SeriesInstanceUID (meaningless value, but reproducible)
+
     """
-    import time
     import calendar
 
     dicom = dcm.read_file(dicom_list[0], stop_before_pixels=True, force=True)
-    dcm_date = dicom.SeriesDate  # YYYYMMDD
-    dcm_time = dicom.SeriesTime  # HHMMSS.MICROSEC
-    dicom_time_str = dcm_date + dcm_time.split('.', 1)[0]  # YYYYMMDDHHMMSS
-    # convert to epoch
-    return calendar.timegm(time.strptime(dicom_time_str, '%Y%m%d%H%M%S'))
+    if (dicom_datetime := get_datetime_from_dcm(dicom)):
+        return calendar.timegm(dicom_datetime.timetuple())
+    else:
+        logging.warning("unable to get real timestamp from series. returning arbitrary time based on hash of SeriesInstanceUID")
+        return hash(dicom.SeriesInstanceUID)
```

Comment:
here -- I wonder if we could avoid `hash` but take the last portion of UID (or just even entire UID with `'.'` removed) -- and then take remainder of dividing by something like 3600*24*265*20 seconds (20 years). Wouldn't those then be sortable in original DICOMs order ? (need to check if SeriesInstanceUID is time based, i.e. later ones would have larger number)  That would address my concern on sorting I had in https://github.com/nipy/heudiconv/issues/612#issuecomment-1338524034

Reply:
Following up on sorting at #612 . But, from what I'm seeing, although the UID can sometimes contain easily extractable information like what you're suggesting, it is explicitly advised against trying to parse the different components in general

https://dicom.nema.org/dicom/2013/output/chtml/part05/chapter_9.html

> Although a specific implementation may choose some particular structure for its generated UIDs, it should never assume that a UID carries any semantics. Thus, a UID shall not be "parsed" to find a particular value or component. Component definition (for the suffix) is implementation specific and may change as long as uniqueness is maintained. Parsing UIDs may jeopardize the ability to inter-operate as implementations evolve.

https://dicom.nema.org/dicom/2013/output/chtml/part05/chapter_B.html

> Because of the flexibility allowed by the DICOM Standard in creating organizationally derived UIDs, implementations should not depend on any assumed structure of UIDs and should not attempt to parse UIDs to extract the semantics of some of its components.